semaphore_units are not semaphore-move safe

semaphore_unit currently contain a naked pointer to the semaphore they got the units from.
See https://github.com/scylladb/seastar/blob/0ec54eab1cf5c2dccb9a1403804128fe21143a21/include/seastar/core/semaphore.hh#L433

But there's nothing to update that pointer if the semaphore is moved.

This can be demonstrated by the following unit test:
```c++
SEASTAR_THREAD_TEST_CASE(test_semaphore_move_with_outstanding_units) {
    auto sem = semaphore(1);
    auto units = std::make_unique<semaphore_units<>>(get_units(sem, 1).get());
    BOOST_REQUIRE_EQUAL(sem.current(), 0);
    auto sem1 = std::move(sem);
    BOOST_REQUIRE_EQUAL(sem1.current(), 0);
    units.reset();
    BOOST_REQUIRE_EQUAL(sem1.current(), 1);
}
```

Currently results in:
```
/home/bhalevy/dev/seastar/tests/unit/semaphore_test.cc(442): fatal error: in "test_semaphore_move_with_outstanding_units": critical check sem1.current() == 1 has failed [0 != 1]
```

> In C++17 you don't need a move constructor for mandated copy elision: https://godbolt.org/z/PcGvefd5z

When I delete the move ctor seastar doesn't build:
```
FAILED: CMakeFiles/seastar.dir/src/core/smp.cc.o 
/usr/local/bin/clang++ -DFMT_SHARED -DSEASTAR_API_LEVEL=6 -DSEASTAR_BROKEN_SOURCE_LOCATION -DSEASTAR_DEFERRED_ACTION_REQUIRE_NOEXCEPT -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_ASAN_FIBER_SUPPORT -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_LZ4_COMPRESS_DEFAULT -DSEASTAR_HAVE_NUMA -DSEASTAR_HAVE_URING -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_THREAD_STACK_GUARDS -DSEASTAR_TYPE_ERASE_MORE -I/home/bhalevy/dev/seastar/include -I/home/bhalevy/dev/seastar/build/dev/gen/include -I/home/bhalevy/dev/seastar/src -I/home/bhalevy/dev/seastar/build/dev/gen/src -O2 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Wno-error=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fvisibility=hidden -UNDEBUG -Wall -Werror -Wno-array-bounds -Wno-error=deprecated-declarations -gz -std=gnu++20 -MD -MT CMakeFiles/seastar.dir/src/core/smp.cc.o -MF CMakeFiles/seastar.dir/src/core/smp.cc.o.d -o CMakeFiles/seastar.dir/src/core/smp.cc.o -c /home/bhalevy/dev/seastar/src/core/smp.cc
In file included from /home/bhalevy/dev/seastar/src/core/smp.cc:22:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/smp.hh:24:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/future.hh:24:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/task.hh:24:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/memory:63:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_algobase.h:67:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_iterator.h:85:
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_construct.h:115:4: error: no matching function for call to 'construct_at'
          std::construct_at(__p, std::forward<_Args>(__args)...);
          ^~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:120:11: note: in instantiation of function template specialization 'std::_Construct<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>>' requested here
            std::_Construct(std::__addressof(*__cur), *__first);
                 ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:370:14: note: in instantiation of function template specialization 'std::__do_uninit_copy<std::move_iterator<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>' requested here
        return std::__do_uninit_copy(__first, __last, __result);
                    ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_vector.h:1617:13: note: in instantiation of function template specialization 'std::__uninitialized_copy_a<std::move_iterator<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>>' requested here
              std::__uninitialized_copy_a(__first, __last, __result,
                   ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/vector.tcc:86:16: note: in instantiation of function template specialization 'std::vector<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>>::_M_allocate_and_copy<std::move_iterator<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>>' requested here
              __tmp = _M_allocate_and_copy(__n,
                      ^
/home/bhalevy/dev/seastar/src/core/smp.cc:75:48: note: in instantiation of member function 'std::vector<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>>::reserve' requested here
                smp_service_groups[id].clients.reserve(smp::count); // may throw
                                               ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_construct.h:94:5: note: candidate template ignored: substitution failure [with _Tp = seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>, _Args = <seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>>]: call to deleted constructor of 'seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>'
    construct_at(_Tp* __location, _Args&&... __args)
    ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_construct.h:119:25: error: call to deleted constructor of 'seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>'
      ::new((void*)__p) _Tp(std::forward<_Args>(__args)...);
                        ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bhalevy/dev/seastar/include/seastar/core/semaphore.hh:248:5: note: 'basic_semaphore' has been explicitly marked deleted here
    basic_semaphore(basic_semaphore&&) = delete;
    ^
In file included from /home/bhalevy/dev/seastar/src/core/smp.cc:22:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/smp.hh:24:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/future.hh:24:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/task.hh:24:
In file included from /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/memory:66:
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:90:7: error: static assertion failed due to requirement 'is_constructible<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> &&>::value': result type must be constructible from input type
      static_assert(is_constructible<_ValueType, _Tp>::value,
      ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:182:4: note: in instantiation of function template specialization 'std::__check_constructible<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> &&>' requested here
        = _GLIBCXX_USE_ASSIGN_FOR_INIT(_ValueType2, _From);
          ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:101:13: note: expanded from macro '_GLIBCXX_USE_ASSIGN_FOR_INIT'
    && std::__check_constructible<T, U>()
            ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/stl_uninitialized.h:372:19: note: in instantiation of function template specialization 'std::uninitialized_copy<std::move_iterator<seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>, seastar::basic_semaphore<seastar::named_semaphore_exception_factory, seastar::lowres_clock> *>' requested here
      return std::uninitialized_copy(__first, __last, __result);
                  ^
3 errors generated.
[28/219] Building CXX object CMakeFiles/seastar.dir/src/core/reactor.cc.o
FAILED: CMakeFiles/seastar.dir/src/core/reactor.cc.o 
/usr/local/bin/clang++ -DFMT_SHARED -DSEASTAR_API_LEVEL=6 -DSEASTAR_BROKEN_SOURCE_LOCATION -DSEASTAR_DEFERRED_ACTION_REQUIRE_NOEXCEPT -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_ASAN_FIBER_SUPPORT -DSEASTAR_HAVE_HWLOC -DSEASTAR_HAVE_LZ4_COMPRESS_DEFAULT -DSEASTAR_HAVE_NUMA -DSEASTAR_HAVE_URING -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_THREAD_STACK_GUARDS -DSEASTAR_TYPE_ERASE_MORE -I/home/bhalevy/dev/seastar/include -I/home/bhalevy/dev/seastar/build/dev/gen/include -I/home/bhalevy/dev/seastar/src -I/home/bhalevy/dev/seastar/build/dev/gen/src -O2 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Wno-error=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fvisibility=hidden -UNDEBUG -Wall -Werror -Wno-array-bounds -Wno-error=deprecated-declarations -gz -std=gnu++20 -MD -MT CMakeFiles/seastar.dir/src/core/reactor.cc.o -MF CMakeFiles/seastar.dir/src/core/reactor.cc.o.d -o CMakeFiles/seastar.dir/src/core/reactor.cc.o -c /home/bhalevy/dev/seastar/src/core/reactor.cc
In file included from /home/bhalevy/dev/seastar/src/core/reactor.cc:35:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/reactor.hh:25:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/iostream.hh:440:
In file included from /home/bhalevy/dev/seastar/include/seastar/core/iostream-impl.hh:26:
/home/bhalevy/dev/seastar/include/seastar/core/do_with.hh:42:43: error: no matching constructor for initialization of 'std::tuple<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>'
    explicit do_with_state(T&&... args) : _held(std::forward<T>(args)...) {}
                                          ^     ~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/unique_ptr.h:1065:34: note: in instantiation of function template specialization 'seastar::internal::do_with_state<std::tuple<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>, seastar::future<>>::do_with_state<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>' requested here
    { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                                 ^
/home/bhalevy/dev/seastar/include/seastar/core/do_with.hh:93:25: note: in instantiation of function template specialization 'std::make_unique<seastar::internal::do_with_state<std::tuple<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>, seastar::future<>>, seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>' requested here
            return std::make_unique<internal::do_with_state<value_tuple, ret_type>>(std::forward<decltype(x)>(x)...);
                        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:754:17: note: candidate constructor not viable: no known conversion from 'seastar::basic_semaphore<seastar::semaphore_default_exception_factory>' to 'const std::tuple<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>' for 1st argument
      constexpr tuple(const tuple&) = default;
                ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:756:17: note: candidate constructor not viable: no known conversion from 'seastar::basic_semaphore<seastar::semaphore_default_exception_factory>' to 'std::tuple<seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>' for 1st argument
      constexpr tuple(tuple&&) = default;
                ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:727:2: note: candidate template ignored: requirement '_TupleConstraints<true, seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::steady_clock>>::__is_implicitly_constructible()' was not satisfied [with _NotEmpty = true]
        tuple(const _Elements&... __elements)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:734:2: note: candidate template ignored: requirement '_TupleConstraints<true, seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::steady_clock>>::__is_explicitly_constructible()' was not satisfied [with _NotEmpty = true]
        tuple(const _Elements&... __elements)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:742:2: note: candidate template ignored: requirement '_TupleConstraints<true, seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::steady_clock>>::__is_implicitly_constructible()' was not satisfied [with _UElements = <seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>, _Valid = true]
        tuple(_UElements&&... __elements)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:750:2: note: candidate template ignored: requirement '_TupleConstraints<true, seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::steady_clock>>::__is_explicitly_constructible()' was not satisfied [with _UElements = <seastar::basic_semaphore<seastar::semaphore_default_exception_factory>>, _Valid = true]
        tuple(_UElements&&... __elements)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:763:2: note: candidate template ignored: could not match 'tuple' against 'basic_semaphore'
        tuple(const tuple<_UElements...>& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:773:2: note: candidate template ignored: could not match 'tuple' against 'basic_semaphore'
        tuple(const tuple<_UElements...>& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:783:2: note: candidate template ignored: could not match 'tuple' against 'basic_semaphore'
        tuple(tuple<_UElements...>&& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:792:2: note: candidate template ignored: could not match 'tuple' against 'basic_semaphore'
        tuple(tuple<_UElements...>&& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:713:2: note: candidate constructor template not viable: requires 0 arguments, but 1 was provided
        tuple()
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:720:2: note: candidate constructor template not viable: requires 0 arguments, but 1 was provided
        tuple()
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:801:2: note: candidate constructor template not viable: requires 2 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:807:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:815:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:823:2: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:833:2: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:840:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a, const tuple& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:845:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a, tuple&& __in)
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:853:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:865:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:876:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
/usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/tuple:888:2: note: candidate constructor template not viable: requires 3 arguments, but 1 was provided
        tuple(allocator_arg_t __tag, const _Alloc& __a,
        ^
1 error generated.
```

--- a/include/seastar/core/semaphore.hh
+++ b/include/seastar/core/semaphore.hh
@@ -132,9 +132,6 @@ struct named_semaphore_exception_factory {
     named_semaphore_aborted aborted() const noexcept;
 };
 
-template<typename ExceptionFactory, typename Clock>
-class semaphore_units;
-
 /// \brief Counted resource guard.
 ///
 /// This is a standard computer science semaphore, adapted
@@ -164,24 +161,6 @@ public:
 private:
     ssize_t _count;
     std::exception_ptr _ex;
-#ifdef SEASTAR_SEMAPHORE_DEBUG
-    size_t _outstanding_units = 0;
-
-    void assert_not_held() const noexcept {
-        assert(!_outstanding_units && "semaphore moved with outstanding units");
-    }
-    void add_outstanding_units(size_t n) {
-        _outstanding_units += n;
-    }
-    void sub_outstanding_units(size_t n) {
-        _outstanding_units -= n;
-    }
-#else   // SEASTAR_SEMAPHORE_DEBUG
-    void assert_not_held() const noexcept {}
-    void add_outstanding_units(size_t) {}
-    void sub_outstanding_units(size_t) {}
-#endif  // SEASTAR_SEMAPHORE_DEBUG
-
     struct entry {
         promise<> pr;
         size_t nr;
@@ -213,14 +192,41 @@ private:
         }
     };
     internal::abortable_fifo<entry, expiry_handler> _wait_list;
+
+#ifdef SEASTAR_SEMAPHORE_DEBUG
+    struct used_flag {
+        // set to true from the wait path
+        // prevents the semaphore from being moved or move-reassigned when _used
+        bool _used = false;
+
+        used_flag() = default;
+        used_flag(used_flag&& o) noexcept {
+            assert(!_used && "semaphore cannot be moved after it has been used");
+        }
+        used_flag& operator=(used_flag&& o) noexcept {
+            if (this != &o) {
+                assert(!_used && !o._used && "semaphore cannot be moved after it has been used");
+            }
+            return *this;
+        }
+        void use() noexcept {
+            _used = true;
+        }
+    };
+#else
+    struct used_flag {
+        void use() noexcept {}
+    };
+#endif
+
+    [[no_unique_address]] used_flag _used;
+
     bool has_available_units(size_t nr) const noexcept {
         return _count >= 0 && (static_cast<size_t>(_count) >= nr);
     }
     bool may_proceed(size_t nr) const noexcept {
         return has_available_units(nr) && _wait_list.empty();
     }
-
-    friend class semaphore_units<ExceptionFactory, Clock>;
 public:
     /// Returns the maximum number of units the semaphore counter can hold
     static constexpr size_t max_counter() noexcept {
@@ -244,27 +250,6 @@ public:
     {
         static_assert(std::is_nothrow_move_constructible_v<expiry_handler>);
     }
-
-    basic_semaphore(const basic_semaphore&) = delete;
-    basic_semaphore(basic_semaphore&& o) noexcept
-        : _count(std::exchange(o._count, 0))
-        , _ex(std::move(o._ex))
-        , _wait_list(std::move(o._wait_list))
-    {
-        o.assert_not_held();
-    }
-
-    basic_semaphore& operator=(basic_semaphore&& o) noexcept {
-        if (this != &o) {
-            o.assert_not_held();
-            assert_not_held();
-            _count = std::exchange(o._count, 0);
-            _ex = std::move(o._ex);
-            _wait_list = std::move(o._wait_list);
-        }
-        return *this;
-    }
-
     /// Waits until at least a specific number of units are available in the
     /// counter, and reduces the counter by that amount of units.
     ///
@@ -292,6 +277,7 @@ public:
     ///         \ref semaphore_timed_out exception.  If the semaphore was
     ///         \ref broken(), may contain a \ref broken_semaphore exception.
     future<> wait(time_point timeout, size_t nr = 1) noexcept {
+        _used.use();
         if (may_proceed(nr)) {
             _count -= nr;
             return make_ready_future<>();
@@ -327,6 +313,7 @@ public:
     ///         \ref semaphore_aborted exception.  If the semaphore was
     ///         \ref broken(), may contain a \ref broken_semaphore exception.
     future<> wait(abort_source& as, size_t nr = 1) noexcept {
+        _used.use();
         if (may_proceed(nr)) {
             _count -= nr;
             return make_ready_future<>();
@@ -391,6 +378,7 @@ public:
     ///
     /// \param nr Amount of units to consume (default 1).
     void consume(size_t nr = 1) noexcept {
+        _used.use();
         if (_ex) {
             return;
         }
@@ -408,6 +396,7 @@ public:
     /// \param nr number of units to reduce the counter by (default 1).
     /// \return `true` if the counter had sufficient units, and was decremented.
     bool try_wait(size_t nr = 1) noexcept {
+        _used.use();
         if (may_proceed(nr)) {
             _count -= nr;
             return true;
@@ -484,11 +473,9 @@ class semaphore_units {
     basic_semaphore<ExceptionFactory, Clock>* _sem;
     size_t _n;
 
-    semaphore_units(basic_semaphore<ExceptionFactory, Clock>* sem, size_t n) noexcept : _sem(sem), _n(n) {
-        _sem->add_outstanding_units(n);
-    }
+    semaphore_units(basic_semaphore<ExceptionFactory, Clock>* sem, size_t n) noexcept : _sem(sem), _n(n) {}
 public:
-    semaphore_units() noexcept : _sem(nullptr), _n(0) {}
+    semaphore_units() noexcept : semaphore_units(nullptr, 0) {}
     semaphore_units(basic_semaphore<ExceptionFactory, Clock>& sem, size_t n) noexcept : semaphore_units(&sem, n) {}
     semaphore_units(semaphore_units&& o) noexcept : _sem(o._sem), _n(std::exchange(o._n, 0)) {
     }
@@ -513,14 +500,12 @@ public:
             throw std::invalid_argument("Cannot take more units than those protected by the semaphore");
         }
         _n -= units;
-        _sem->sub_outstanding_units(units);
         _sem->signal(units);
         return _n;
     }
     /// Return ownership of all units. The semaphore will be signaled by the number of units returned.
     void return_all() noexcept {
         if (_n) {
-            _sem->sub_outstanding_units(_n);
             _sem->signal(_n);
             _n = 0;
         }
@@ -529,7 +514,6 @@ public:
     ///
     /// \return the number of units held
     size_t release() noexcept {
-        _sem->sub_outstanding_units(_n);
         return std::exchange(_n, 0);
     }
     /// Splits this instance into a \ref semaphore_units object holding the specified amount of units.
@@ -545,11 +529,6 @@ public:
             throw std::invalid_argument("Cannot take more units than those protected by the semaphore");
         }
         _n -= units;
-        // `units` are split into a new `semaphore_units` object.
-        // since none are returned to the semaphore we subtract the
-        // outstanding units here to balance their eventual addition by the
-        // `semaphore_units` ctor.
-        _sem->sub_outstanding_units(units);
         return semaphore_units(_sem, units);
     }
     /// The inverse of split(), in which the units held by the specified \ref semaphore_units
@@ -559,9 +538,7 @@ public:
     /// \return the updated semaphore_units object
     void adopt(semaphore_units&& other) noexcept {
         assert(other._sem == _sem);
-        auto o = other.release();
-        _sem->add_outstanding_units(o);
-        _n += o;
+        _n += other.release();
     }
 
     /// Returns the number of units held

